packages/rum/src/domain/record/privacy.ts

```diff
@@ -1,91 +1,356 @@
-
-// Returns true if the given DOM node should be hidden. Ancestors
-// are not checked.
-export function nodeShouldBeHidden(node: Node): boolean {
-  return (
-    isElement(node) &&
-    (node.getAttribute(PRIVACY_ATTR_NAME) === PRIVACY_ATTR_VALUE_HIDDEN ||
-      node.classList.contains(PRIVACY_CLASS_HIDDEN))
-  )
+import { shouldIgnoreElement } from './serialize'
+
+const TEXT_MASKING_CHAR = '᙮'
+const MIN_LEN_TO_MASK = 80
+const WHITESPACE_TEST = /^\s$/
+
+export function getInitialPrivacyLevel(): NodePrivacyLevelInternal {
+  return NodePrivacyLevelInternal.ALLOW
 }
 
-// Returns true if the given DOM node should be hidden, recursively
-// checking its ancestors.
-export function nodeOrAncestorsShouldBeHidden(node: Node | null): boolean {
-  if (!node) {
-    return false
+/**
+ * PUBLIC: Resolves the internal privacy level and remaps to level to the format
+ * exposed to the general codebase  (allow/mask/hidden/ignore) because
+ * NOT_SET/UNKNOWN/*-SEALED/MASK_FORMS_ONLY are not dev friendly and need not be handled by devs
+ */
+export function getNodePrivacyLevel(node: Node, parentNodePrivacyLevel?: NodePrivacyLevelInternal): NodePrivacyLevel {
+  const privacyLevel = getInternalNodePrivacyLevel(node, parentNodePrivacyLevel)
+  return remapInternalPrivacyLevels(node, privacyLevel)
+}
+
+/**
+ * Remap the internal privacy levels to general use privacy levels
+ */
+export function remapInternalPrivacyLevels(
+  node: Node,
+  nodePrivacyLevelInternal: NodePrivacyLevelInternal
+): NodePrivacyLevel {
+  switch (nodePrivacyLevelInternal) {
+    // Pass through levels
+    case NodePrivacyLevelInternal.ALLOW:
+      return NodePrivacyLevel.ALLOW
+    case NodePrivacyLevelInternal.MASK:
+      return NodePrivacyLevel.MASK
+    case NodePrivacyLevelInternal.HIDDEN:
+      return NodePrivacyLevel.HIDDEN
+    case NodePrivacyLevelInternal.IGNORE:
+      return NodePrivacyLevel.IGNORE
+    // Conditional levels
+    case NodePrivacyLevelInternal.MASK_FORMS_ONLY:
+      return isFormElement(node) ? NodePrivacyLevel.MASK : NodePrivacyLevel.ALLOW
+    default:
+    // Edgecase handling: TODO: REVIEW: `hide`, `mask`, or `allow`?
+    case NodePrivacyLevelInternal.UNKNOWN:
+    case NodePrivacyLevelInternal.NOT_SET:
+      return NodePrivacyLevel.ALLOW
   }
+}
 
-  if (nodeShouldBeHidden(node)) {
-    return true
+/**
+ * INTERNAL FUNC: Get privacy level without remapping (or setting cache)
+ * This function may be explicitly used when passing internal privacy levels to
+ * child nodes for performance reasons, otherwise you should use `getNodePrivacyLevel`
+ */
+export function getInternalNodePrivacyLevel(
+  node: Node,
+  parentNodePrivacyLevel?: NodePrivacyLevelInternal
+): NodePrivacyLevelInternal {
+  const isElementNode = isElement(node)
+
+  // Only Elements have tags directly applied
+  if (node.parentElement && !isElementNode) {
+    return parentNodePrivacyLevel || getInternalNodePrivacyLevel(node.parentElement)
+  }
```

Comment:
Checking here let's us skip `derivePrivacyLevelGivenParent()`.   Also currently `getNodeSelfPrivacyLevel` doesn't really handle non-element checks because it is checked here. Could be reorganised, but I think this is better to catch earlier rather than on a somewhat "warmer path" like `getNodeSelfPrivacyLevel`.